packages/block-editor/src/hooks/border.js

```diff
@@ -1,53 +1,169 @@
+	const { borderColor, sideBorderColors, style } = props.attributes;
+	return (
+		isDefinedBorder( style?.border ) ||
+		!! borderColor ||
+		!! sideBorderColors
+	);
+};
+
+// The border color, style, and width are omitted so they get undefined. The
+// border radius is separate and must retain its selection.
+const resetBorder = ( { attributes = {}, setAttributes } ) => {
+	const { style } = attributes;
+	setAttributes( {
+		borderColor: undefined,
+		sideBorderColors: undefined,
+		style: {
+			...style,
+			border: cleanEmptyObject( {
+				radius: style?.border?.radius,
+			} ),
+		},
+	} );
+};
+
+const resetBorderFilter = ( newAttributes ) => ( {
+	...newAttributes,
+	borderColor: undefined,
+	sideBorderColors: undefined,
+	style: {
+		...newAttributes.style,
+		border: {
+			radius: newAttributes.style?.border?.radius,
+		},
+	},
+} );
+
+const getColorByProperty = ( colors, property, value ) => {
+	let matchedColor;
+
+	colors.some( ( origin ) =>
+		origin.colors.some( ( color ) => {
+			if ( color[ property ] === value ) {
+				matchedColor = color;
+				return true;
+			}
+
+			return false;
+		} )
+	);
+
+	return matchedColor;
+};
+
+export const getMultiOriginColor = ( { colors, namedColor, customColor } ) => {
+	// Search each origin (default, theme, or user) for matching color by name.
+	if ( namedColor ) {
+		const colorObject = getColorByProperty( colors, 'slug', namedColor );
+		if ( colorObject ) {
+			return colorObject;
+		}
+	}
+
+	// Skip if no custom color or matching named color.
+	if ( ! customColor ) {
+		return { color: undefined };
+	}
+
+	// Attempt to find color via custom color value or build new object.
+	const colorObject = getColorByProperty( colors, 'color', customColor );
+	return colorObject ? colorObject : { color: customColor };
+};
+
+const getBorderObject = ( attributes, colors ) => {
+	const { borderColor, sideBorderColors, style } = attributes;
+	const { border: borderStyles } = style || {};
+
+	// If we have a named color for a flat border. Fetch that color object and
+	// apply that color's value to the color property within the style object.
+	if ( borderColor ) {
+		const { color } = getMultiOriginColor( {
+			colors,
+			namedColor: borderColor,
+		} );
+
+		return color ? { ...borderStyles, color } : borderStyles;
+	}
+
+	// If we have named colors for the individual side borders, retrieve their
+	// related color objects and apply the real color values to the split
+	// border objects.
+	if ( sideBorderColors ) {
+		const hydratedBorderStyles = { ...borderStyles };
+
+		Object.entries( sideBorderColors ).forEach(
+			( [ side, namedColor ] ) => {
+				const { color } = getMultiOriginColor( { colors, namedColor } );
+
+				if ( color ) {
+					hydratedBorderStyles[ side ] = {
+						...hydratedBorderStyles[ side ],
+						color,
+					};
+				}
+			}
+		);
+
+		return hydratedBorderStyles;
+	}
+
+	// No named colors selected all color values if any should already be in
```

Comment:
Tiny nit: I wasn't sure at first what this sentence meant, so I wondered if this could clear it up a little:

```suggestion
	// No named colors selected. All color values (if any) should already be in
```